lib/hanami/action/validatable.rb

```diff
@@ -95,12 +95,101 @@ module ClassMethods
         #     end
         #   end
         #
+        # @api public
         # @since 0.3.0
+        def params(klass = nil, &block)
+          if klass.nil?
+            klass = const_set(PARAMS_CLASS_NAME, Class.new(Params))
+            klass.params(&block)
+          end
+
+          @params_class = klass
+        end
+
+        # Defines a validation contract for the params passed to {Hanami::Action#call}.
+        #
+        # This feature isn't mandatory, but is highly recommended for secure handling of params:
+        # because params come from an untrusted source, it's good practice to filter these to only
+        # the keys and types required for your action's use case.
+        #
+        # The given block is evaluated inside a `Dry::Validation::Contract` class. This allows you
+        # to use all features of dry-validation contracts
+        #
+        # The resulting contract becomes part of a dedicated params class for the action, inheriting
+        # from {Hanami::Action::Params}.
+        #
+        # Instead of defining the params validation contract inline, you can alternatively provide a
+        # concrete params class, which should inherit from {Hanami::Action::Params}.
+        #
+        # @param klass [Class,nil] a Hanami::Action::Params subclass
+        # @param block [Proc] the params schema definition
+        #
+        # @return void
+        #
+        # @see #params
+        # @see Hanami::Action::Params
+        # @see https://dry-rb.org/gems/dry-validation/
+        #
+        # @example Inline definition
+        #   require "hanami/controller"
+        #
+        #   class Signup < Hanami::Action
+        #     contract do
+        #       params do
+        #         required(:first_name)
+        #         required(:last_name)
+        #         required(:email)
+        #       end
+        #
+        #       rule(:email) do
+        #         # custom rule logic here
+        #       end
+        #     end
+        #
+        #     def handle(req, *)
+        #       puts req.params.class            # => Signup::Params
+        #       puts req.params.class.superclass # => Hanami::Action::Params
+        #
+        #       puts req.params[:first_name]     # => "Luca"
+        #       puts req.params[:admin]          # => nil
+        #     end
+        #   end
+        #
+        # @example Concrete class
+        #   require "hanami/controller"
+        #
+        #   class SignupParams < Hanami::Action::Params
```

Comment:
I'm conflicted about this decision. The reason why I would most likely want to use a concrete class for this is to share validation rules between multiple things.

By basing this on `Hanami::Action::Params`, this requires a request env to initialize, which limits its usefulness to controllers only. This excludes non-request sharing, such as Operations or CLI tools.

I wish contracts were composable. If I could merge a preexisting contract into the action validator, this would be moot.

Another way you can share logic is via [macros](https://dry-rb.org/gems/dry-validation/1.10/macros/). My preference would be to define an abstract class to contain these macros because I may want to use different ones for different use-cases. But since you hard-code `Dry::Validation::Contract` as the base class here, I can only share macros globally. I don't know if that's a problem.

On the other hand, perhaps its reasonable that HTML-boundary validations are their own thing.

Reply:
@alassek Thanks for raising this! I agree with you here. If I was building this feature from scratch today, I wouldn't have things run through these `Hanami::Action::Params` intermediary classes, and instead have dry-validation contracts be provided directly.

I'd describe this current approach as a vestige of the 1.x-era of Hanami design, where it was much more heavy-handed in "owning the APIs" rather than the lighter touch approach we take for our integration with the dry-rb gems today. It's also worth mentioning that back in those times, dry-validation might have seemed like a slightly less stable target, so the framework builders might've wanted this extra layer as "insulation."

In terms of why I implemented it this way here: I wanted a small, minimally-disruptive  change as a first step, something I could merge confidently and use as a launching point for any next steps. I didn't say it out loud, but I did want us to get to the point of supporting dry-validation contracts directly, but I wasn't sure about when would be ideal for that.

Your question motivates me to have a quick run at it as a follow-up and see what might be possible :)

As an additional measure, I'm going to mark the new `Hanami::API::Params.contract` method here as `@api private`, as part of a "soft deprecation" approach to this direct use of the Params class. I'd much rather have users specify their validations at the action level only, and use contracts directly as a way to reuse it. Let's see how it goes!

Reply:
I've started this work in #454. I think I have a pretty clear path to getting it done, too. I'll leave another update over in that PR once it's ready!

Reply:
#454 is ready now :)